Chore/1333 replace terraform cloud

joshgamache opened the pull request:
Addresses #1333 - [Project link](https://github.com/orgs/bcgov/projects/123/views/2?filterQuery=1333&pane=issue&itemId=49863829)
Adds Terraform scripts and Helm charts to run them. Resource states migrated from Terraform Cloud to a Google Cloud Storage bucket.

## Changes 🚧

- Added Terraform module, imported from cas-shelf
- Added Helm templates to handle:
  - Terraform module loading to configMaps
  - OpenShift service account for use by Terraform
  - Terraform apply job

## To test 🔬

- The Helm charts are built into the automation around deploys via ShipIt. Once a deploy has been run, you can check the job `Terraform-apply` in OpenShift to see if the run was successful.

## Notes📝 
- 🔒 Terraform Cloud state has been locked to enable migration of state to state buckets
- The script has been run and resources have been transitioned from Terraform Cloud to GCP buckets holding state for each namespace (cif-dev, cif-test, cif-prod).  
- Resources have been migrated rather than recreated, only the state "owning" the resource changes. 
- The resources `kubernetes_secret.secret_sa` have had Terraform changes, but only the addition of a default value (consistent with our use of them) due to a new provider version. 
- The script used for migration (/chart/cas-cif/terraform/tf-migration.sh) is kept here for history. It will be removed by https://github.com/bcgov/cas-pipeline/issues/83

pbastia commented on chart/cas-cif/templates/jobs/terraform-service-account.yaml:
We typically use only one file per resource in our helm chart, this makes it more readable especially for people who don't know how things are setup. Unless there is a reason for having them together?

joshgamache replied:
Since they were all directly related, I stitched them together for convenience. But there's no reason they need to be that way. I've addressed that in 8a55f15b.

pbastia commented on chart/cas-cif/templates/jobs/terraform-apply.yaml:
```diff
@@ -0,0 +1,73 @@
+            - name: GOOGLE_APPLICATION_CREDENTIALS
+              value: "/etc/gcp/credentials.json"
+          # Terraform was having an issue pulling kubernetes_host in as a TF_VAR, so we add it as a attribute to the command
+          command:
+            [
+              "/bin/sh",
+              "-c",
+              "cp -r /terraform/. /working && cd working && export TF_VAR_kubernetes_token=$( cat /var/run/secrets/kubernetes.io/serviceaccount/token ) && terraform init -backend-config=/etc/tf/gcs.tfbackend && terraform apply -var=\"kubernetes_host=$kubernetes_host\" -auto-approve",
+            ]
```
We might want to add a `set -euxo pipefail` to help debugging if the command fails
Helm sucks, and there's a slightly easier syntax to pass in bash scripts:

```suggestion
          command:
              - /bin/sh
              - -c
              - |
                  set -euo pipefail;
                  cp -r /terraform/. /working;
                  cd working;
                  export TF_VAR_kubernetes_token=$( cat /var/run/secrets/kubernetes.io/serviceaccount/token );
                  terraform init -backend-config=/etc/tf/gcs.tfbackend;
                  terraform apply -var=\"kubernetes_host=$kubernetes_host\" -auto-approve";
```

joshgamache replied:
Thanks for the suggestion. I knew the long syntax wasn't the easiest to read, this makes it _much_ easier to understand! Changed with d068385a.

pbastia commented on chart/cas-cif/templates/jobs/terraform-apply.yaml:
```diff
@@ -0,0 +1,73 @@
+              name: terraform-backend-config-volume
+              readOnly: True
+            - name: tf-cache
+              mountPath: /working
```
Because of `emptyDir` the working directory will be deleted as soon as the job pod completes. Should we name this mount `tf-working-dir` or something like that? Since it doesn't cache anything

pbastia reviewed (changes requested):
Nice work! A few comments about bash formatting, looks good otherwise.
I'd suggest also:
- moving `tf-migration.sh` and `migration.example.tfvars` out of the helm chart, I don't think there is a to have these in there?
- move all the terraform yaml templates to `chart/cas-cif/templates/terraform` instead of the `jobs` folder

joshgamache commented:
> * moving `tf-migration.sh` and `migration.example.tfvars` out of the helm chart, I don't think there is a to have these in there?

I cleaned up the directories with the suggestions you made. Especially with the service account related file split, it keeps them consistently together. 

I removed the migration script entirely, as it is unneeded now that the migrations have been run.

pbastia reviewed (approved):
Slick!

mikevespi reviewed (approved):
Awesome stuff! Leaving it Approved but not merged as per discussion today about merging with eyes on. Aside from that let's make sure we get a tech-debt card to create custom GCP role instead of their generic admin, they over permit.

Great stuff!

mikevespi commented on chart/cas-cif/.helmignore:
```diff
@@ -21,3 +21,11 @@
 .idea/
 *.tmproj
 .vscode/
+
```
Can we add the below to the .gitignore as well :)

joshgamache replied:
Added in 81e56976e87a31cf74ca2dd3c1d6d820efd2295e
